metpy/calc/tools.py

```diff
@@ -330,3 +332,185 @@ def log_interp(x, xp, fp, **kwargs):
+
+    This function will subset an upper air dataset to contain only the specified layer. The
+    bottom of the layer can be specified with a pressure or height above the surface
+    pressure. The bottom defaults to the surface pressure. The depth of the layer can be
+    specified in terms of pressure or height above the bottom of the layer. If the top and
+    bottom of the layer are not in the data, they are interpolated by default.
+
+    Parameters
+    ----------
+    p : array-like
+        Atmospheric pressure profile
+    *args : array-like
+        Atmospheric variable(s) measured at the given pressures
+    heights: array-like
+        Atmospheric heights corresponding to the given pressures
+    bottom : `pint.Quantity`
+        The bottom of the layer as a pressure or height above the surface pressure
+    depth : `pint.Quantity`
+        The thickness of the layer as a pressure or height above the bottom of the layer
+    interpolate : bool
+        Interpolate the top and bottom points if they are not in the given data
+
+    Returns
+    -------
+    `pint.Quantity, pint.Quantity`
+        The pressure and data variables of the layer
+
+    """
+    # Pop off keyword arguments
+    heights = kwargs.pop('heights', None)
+    bottom = kwargs.pop('bottom', None)
+    depth = kwargs.pop('depth', 100 * units.hPa)
+    interpolate = kwargs.pop('interpolate', True)
+
+    # Make sure pressure and datavar are the same length
+    for datavar in args:
+        if len(p) != len(datavar):
+            raise ValueError('Pressure and data variable must have the same length.')
+
+    # If the bottom is not specified, make it the surface pressure
+    if bottom is None:
+        bottom = p[0]
+
+    bottom_pressure, bottom_height = _get_bound_pressure_height(p, bottom, heights=heights,
+                                                                interpolate=interpolate)
+
+    # Calculate the top if whatever units depth is in
+    if depth.dimensionality == {'[length]': -1.0, '[mass]': 1.0, '[time]': -2.0}:
+        top = bottom_pressure - depth
+    elif depth.dimensionality == {'[length]': 1}:
+        top = bottom_height + depth
+    else:
+        raise ValueError('Depth must be specified in units of length or pressure')
+
+    top_pressure, _ = _get_bound_pressure_height(p, top, heights=heights,
+                                                 interpolate=interpolate)
+
+    ret = []  # returned data variables in layer
+
+    # Ensure pressures are sorted in ascending order
+    sort_inds = np.argsort(p)
+    p = p[sort_inds]
+
+    # Mask based on top and bottom pressure
+    inds = (p <= bottom_pressure) & (p >= top_pressure)
+    p_interp = p[inds]
+
+    # Interpolate pressures at bounds if necessary and sort
+    if interpolate:
+        # If we don't have the bottom or top requested, append them
+        if top_pressure not in p_interp:
```

Comment:
Not sure how I feel about repeated `in`, `append`, and `sort`. Any way to use `searchsorted` and `insert`?